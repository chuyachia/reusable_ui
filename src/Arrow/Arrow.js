import React from "react";
import PropTypes from "prop-types";
const Arrow = props => <i {...props} />;
Arrow.propTypes = {
  className: PropTypes.string,
};
export default Arrow;
