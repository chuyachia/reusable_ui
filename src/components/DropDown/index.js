export {default as DropDownItem} from './DropDownItem';
export {default as DropDownList} from './DropDownItem';