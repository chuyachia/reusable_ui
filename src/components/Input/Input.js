import React from "react";
import PropTypes from "prop-types";

import Button from "../Button";

const Input = ({ resetButton, onReset, onChange, ...props }) => (
  <div {...props}>
    <input onChange={onChange}/>
    {resetButton && (
      <Button onClick={onReset}>
        X
      </Button>
    )}
  </div>
);

Input.propTypes = {
  onChange: PropTypes.func,
  onReset: PropTypes.func,
  resetButton: PropTypes.bool,
  value: PropTypes.string
};

export default Input;
