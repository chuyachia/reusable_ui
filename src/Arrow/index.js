import styled from "@emotion/styled";

import useTheme from "../useTheme";
import Arrow from "./Arrow";

const StyledArrow = styled(Arrow)`
  border-style: solid;
  border-color: ${props => {
    if (props.variant) {
      return useTheme(props, props.variant);
    }
    if (props.disabled) {
      return useTheme(props, "grey");
    }
    return useTheme(props, "baseTextColor");
  }};
  display: inline-block;
  padding: ${props => {
    var size = props.size === undefined ? "3px" : props.size;
    return size;
  }};
  border-width: ${props => {
    var size = props.size === undefined ? "3px" : props.size;
    return "0 " + size + " " + size + " 0";
  }};
  ${props => {
    switch (props.direction) {
      case "right":
        return "transform: rotate(-45deg)";
      case "left":
        return "transform: rotate(135deg)";
      case "up":
        return "transform: translateY(25%) rotate(-135deg)";
      case "down":
        return "transform: translateY(-25%) rotate(45deg)";
    }
  }};
  transition: transform 0.3s;
`;

export default StyledArrow;
