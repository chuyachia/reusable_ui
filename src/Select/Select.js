import React, { useState, useRef, useMemo, useCallback } from "react";
import PropTypes from "prop-types";

import Button from "../Button";
import Input from "../Input/Input";
import Tag from "../Tag";
import Arrow from "../Arrow";
import { DropDownList } from "../DropDown";
import { DropDownItem } from "../DropDown";

const Select = ({
  suggestion,
  multiple,
  selected,
  className,
  onDelete,
  onInput,
  onSelect,
  variant,
  children,
  options,
}) => {
  const [stateOpen, setStateOpen] = useState(false);
  const [textInput, setTextInput] = useState("");
  const [activeItem, setActiveItem] = useState(0);
  const dropDownRef = useRef(null);
  const isSelected = option =>
    selected.findIndex(
      s => s.value === option.value && s.label === option.label
    ) >= 0;
  const inputKeyDown = e => {
    setStateOpen(true);
    var key = e.keyCode || e.charCode;
    if ((key == 8 || key == 46) && textInput.length == 0) {
      e.preventDefault();
      onDelete();
    }
  };
  const deleteSelection = e => {
    onDelete(e.target.getAttribute("value"));
  };
  const selectOption = option => {
    if (multiple && isSelected(option)) {
      onDelete(option.value);
    } else {
      onSelect(option);
    }
    if (onInput) {
      onInput("");
    }
    setTextInput("");
    if (!multiple) setStateOpen(false);
  };
  const handleItemClick = e => {
    if (e.target.nodeName === "LI") {
      const option = {
        value: e.target.getAttribute("value"),
        label: e.target.getAttribute("label"),
      };
      selectOption(option);
    }
  };
  const keyDownOpenDropDown = e => {
    if (e.keyCode === 13) {
      if (!stateOpen) {
        setStateOpen(true);
      } else {
        const option = options[activeItem];
        selectOption(option);
      }
    } else if (e.keyCode === 40) {
      if (!stateOpen) {
        setStateOpen(true);
      } else {
        keyDownChangeActive(e);
      }
    } else if (e.keyCode === 38 && stateOpen) {
      keyDownChangeActive(e);
    }
  };

  const keyDownChangeActive = e => {
    if (e.keyCode === 40 && activeItem < options.length - 1) {
      setActiveItem(activeItem + 1);
      dropDownRef.current.scrollBy(0, 40);
    } else if (e.keyCode === 38 && activeItem > 0) {
      setActiveItem(activeItem - 1);
      dropDownRef.current.scrollBy(0, -40);
    }
  };
  const suggestionInput = suggestion ? (
    <Input
      controlled={true}
      value={textInput}
      onFocus={() => setStateOpen(true)}
      onChange={e => {
        onInput(e.target.value);
        setTextInput(e.target.value);
        setActiveItem(0);
      }}
      onKeyDown={inputKeyDown}
      className="select-input"
    />
  ) : (
    <div className="select-input" />
  );

  const arrowButton = stateOpen ? (
    <Button onClick={() => setStateOpen(!stateOpen)}>
      <Arrow direction="up" />
    </Button>
  ) : (
    <Button onClick={() => setStateOpen(!stateOpen)}>
      <Arrow direction="down" />
    </Button>
  );
  return (
    <div className={className} tabIndex={0} onKeyDown={keyDownOpenDropDown}>
      <div className="select-display">
        {multiple ? (
          <React.Fragment>
            {selected.length > 0 ? (
              selected.map(s => (
                <Tag
                  variant={variant}
                  key={s.value}
                  value={s.value}
                  onClick={deleteSelection}
                >
                  {s.label}
                </Tag>
              ))
            ) : (
              <span>&nbsp;</span>
            )}
          </React.Fragment>
        ) : (
          <React.Fragment>
            {selected && selected.label ? (
              <span>{selected.label}</span>
            ) : (
              <span>&nbsp;</span>
            )}
          </React.Fragment>
        )}
        {suggestionInput}
        {arrowButton}
      </div>
      <DropDownList
        variant={variant}
        open={stateOpen}
        ref={dropDownRef}
        controlled={true}
      >
        {options.map((s, i) => (
          <DropDownItem
            key={s.value}
            {...s}
            disabled={multiple && selected && isSelected(s)}
            onClick={handleItemClick}
            className={i === activeItem && "active"}
            onMouseOver={() => setActiveItem(i)}
          >
            {s.label}
          </DropDownItem>
        ))}
        {options.length === 0 && (
          <DropDownItem disabled={true}>No result found</DropDownItem>
        )}
      </DropDownList>
    </div>
  );
};

Select.propTypes = {
  onInput: PropTypes.func,
  onDelete: PropTypes.func,
  onSelect: PropTypes.func,
  selected: PropTypes.oneOfType([
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string,
    }),
    PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.string,
        label: PropTypes.string,
      })
    ),
  ]),
  variant: PropTypes.string,
  children: PropTypes.node,
  multiple: PropTypes.bool,
  suggestion: PropTypes.bool,
  className: PropTypes.string,
  options: PropTypes.array,
};

export default Select;
