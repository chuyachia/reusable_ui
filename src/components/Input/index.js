import React from "react";
import styled from "@emotion/styled";

import Input from "./Input";
import { ButtonContext } from "../Button";

const StyledInput = styled(Input)`
  display: inline-block;
  white-space: nowrap;
  color: ${props => props.theme.baseTextColor};
  font-weight: ${props => props.theme.fontNormal};
  line-height: ${props => props.theme.baseLineHeight};
  padding: ${props => props.theme.basePadding};
  font-family: ${props => props.theme.fontSans};
  font-size: ${props => props.theme.baseFontSize};
  border-width: ${props => props.theme.borderWidth};
  border-style: solid;
  border-color: ${props => props.theme.primary};
  border-radius: ${props => props.theme.baseRadius};
  & input {
    border-width: 0px;
    padding: 0px;
    line-height: inherit;
    outline: 0;
    font-weight: ${props => props.theme.fontNormal};
    font-size: ${props => props.theme.baseFontSize};
    font-family: ${props => props.theme.fontSans};
  }
`;
export default props => (
  <ButtonContext.Provider value="input">
    <StyledInput {...props} />
  </ButtonContext.Provider>
);
