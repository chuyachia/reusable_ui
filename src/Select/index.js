import React from "react";
import styled from "@emotion/styled";
import PropTypes from "prop-types";

import Select from "./Select";
import { ButtonContext } from "../Button";
import { DropDownListContext } from "../DropDown/DropDownList";
import useTheme from "../useTheme";

const StyledSelect = styled(Select)`
  min-width: ${props => useTheme(props, "inputMinWidth")};
  position: relative;
	display: ${props => (props.inline ? "inline-block" : "block")};
  & div {
    display:inline-block;
  }
  & > div {
    padding: ${props => useTheme(props, "basePadding")};
    font-family: ${props => useTheme(props, "fontSans")};
    font-size: ${props => useTheme(props, "baseFontSize")};
    font-weight: ${props => useTheme(props, "fontNormal")};
    color: ${props => useTheme(props, "baseTextColor")};
    line-height: ${props => useTheme(props, "baseLineHeight")};
    border-width: ${props => useTheme(props, "borderWidth")};
    border-style: solid;
    border-color: ${props => useTheme(props, "primary")};
    border-radius: ${props => useTheme(props, "baseRadius")};
    display:block;
  }
  & input {
    padding: ${props => useTheme(props, "smallPadding")};
    font-family: ${props => useTheme(props, "fontSans")};
    font-size: ${props => useTheme(props, "baseFontSize")}; 
    font-weight: ${props => useTheme(props, "fontNormal")};
    outline:0;
    line-height: ${props => useTheme(props, "baseLineHeight")};
    border:none; 
    width:${props => useTheme(props, "inputMinWidth")}; 
`;

const StyledSelectWContext = props => (
  <ButtonContext.Provider value="select">
    <DropDownListContext.Provider value={props.open}>
      <StyledSelect {...props} />
    </DropDownListContext.Provider>
  </ButtonContext.Provider>
);
StyledSelectWContext.propTypes = {
  open: PropTypes.bool,
};
export default StyledSelectWContext;
