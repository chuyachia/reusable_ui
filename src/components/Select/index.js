import React from "react";
import styled from "@emotion/styled";

import Select from "./Select";
import { ButtonContext } from "../Button";

const StyledSelect =  styled(Select)`
  display:inline-block;
  cursor:pointer;
  & div {
    padding: ${props => props.theme.basePadding};
    font-family: ${props => props.theme.fontSans};
    font-size: ${props => props.theme.baseFontSize};
    font-weight: ${props => props.theme.fontNormal};
    color: ${props => props.theme.baseTextColor};
    line-height: ${props => props.theme.baseLineHeight};
    border-width: ${props => props.theme.borderWidth};
    border-style: solid;
    border-color: ${props => props.theme.primary};
    border-radius: ${props => props.theme.baseRadius};
  }
`;

export default props => (
    <ButtonContext.Provider value="select">
      <StyledSelect {...props} />
    </ButtonContext.Provider>
  );
  