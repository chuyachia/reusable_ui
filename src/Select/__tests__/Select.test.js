import React from "react";

import Select from "../Select";
import Tag from "../../Tag";
import Input from "../../Input/Input";
import Arrow from "../../Arrow";
import { DropDownItem } from "../../DropDown";

const options = [{ value: "a", label: "A" }, { value: "b", label: "B" }, { value: "c", label: "C" }];
describe("Select", () => {
  let wrapper;
  const mockFunc = jest.fn();
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation(init => [init, setState]);
  it("should render single select", () => {
    wrapper = shallow(<Select options={options} selected={{ value: "a", label: "A" }} />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render multiple select", () => {
    wrapper = shallow(<Select options={options} selected={[{ value: "a", label: "A" }, { value: "b", labe: "B" }]} />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should set state open to true when clicked", () => {
    wrapper = shallow(<Select options={options} />);
    let arrow = wrapper.find(Arrow);
    arrow.simulate("click", { stopPropagation: () => null });
    expect(setState).toHaveBeenCalledWith(true);
    jest.clearAllMocks();
  });
  it("should render <Input> when suggestion is true", () => {
    wrapper = shallow(<Select options={options} suggestion />);
    expect(wrapper.find(Input).length).toBe(1);
  });
  it("should call onInput prop and sets state when Input change", () => {
    let event = { target: { value: "aa" } };
    wrapper = shallow(<Select options={options} onInput={mockFunc} suggestion />);
    wrapper.find(Input).simulate("change", event);
    expect(mockFunc.mock.calls.length).toBe(1);
    expect(setState).toHaveBeenCalledWith("aa");
    jest.clearAllMocks();
  });
  it("should call onSelect prop when drop down menu item selected", () => {
    wrapper = shallow(<Select options={options} onSelect={mockFunc} />);
    wrapper.find(DropDownItem).at(0).simulate("click", {
      stopPropagation: () => null,
      target: {
        nodeName: "LI",
        getAttribute: (name) => "a"
      }
    });
    expect(mockFunc.mock.calls.length).toBe(1);
    jest.clearAllMocks();
  });
  it("should call onDelete prop when tag click", () => {
    wrapper = shallow(
      <Select
        options={options}
        selected={[{ value: "a", label: "A" }]}
        onDelete={mockFunc}
        multiple
      />
    );
    let event = { target: { getAttribute: () => "a" } };
    wrapper.find(Tag).simulate("click", event);
    expect(mockFunc.mock.calls.length).toBe(1);
    expect(mockFunc.mock.calls[0][0]).toEqual({ value: "a", label: "a" });
  });
});
