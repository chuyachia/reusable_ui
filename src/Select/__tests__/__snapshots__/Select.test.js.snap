// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render 1`] = `
<div
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="select-display"
  >
    <span>
       
    </span>
    <div
      className="select-input"
    />
    <span
      style={
        Object {
          "float": "right",
        }
      }
    >
      <StyledArrowWContext
        direction="down"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      />
    </span>
  </div>
  <DropDownListWContext
    controlled={true}
    open={false}
  >
    <StyledDropDownItemWContext
      className="active"
      key="a"
      label="A"
      onClick={[Function]}
      onMouseOver={[Function]}
      value="a"
    >
      A
    </StyledDropDownItemWContext>
    <StyledDropDownItemWContext
      className={false}
      key="b"
      label="B"
      onClick={[Function]}
      onMouseOver={[Function]}
      value="b"
    >
      B
    </StyledDropDownItemWContext>
  </DropDownListWContext>
</div>
`;

exports[`Select should render multiple select 1`] = `
<div
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="select-display"
  >
    <span>
       
    </span>
    <div
      className="select-input"
    />
    <span
      style={
        Object {
          "float": "right",
        }
      }
    >
      <StyledArrowWContext
        direction="down"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      />
    </span>
  </div>
  <DropDownListWContext
    controlled={true}
    open={false}
  >
    <StyledDropDownItemWContext
      className="active"
      key="a"
      label="A"
      onClick={[Function]}
      onMouseOver={[Function]}
      value="a"
    >
      A
    </StyledDropDownItemWContext>
    <StyledDropDownItemWContext
      className={false}
      key="b"
      label="B"
      onClick={[Function]}
      onMouseOver={[Function]}
      value="b"
    >
      B
    </StyledDropDownItemWContext>
    <StyledDropDownItemWContext
      className={false}
      key="c"
      label="C"
      onClick={[Function]}
      onMouseOver={[Function]}
      value="c"
    >
      C
    </StyledDropDownItemWContext>
  </DropDownListWContext>
</div>
`;

exports[`Select should render single select 1`] = `
<div
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="select-display"
  >
    <span>
      A
    </span>
    <div
      className="select-input"
    />
    <span
      style={
        Object {
          "float": "right",
        }
      }
    >
      <StyledArrowWContext
        direction="down"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      />
    </span>
  </div>
  <DropDownListWContext
    controlled={true}
    open={false}
  >
    <StyledDropDownItemWContext
      className="active"
      key="a"
      label="A"
      onClick={[Function]}
      onMouseOver={[Function]}
      value="a"
    >
      A
    </StyledDropDownItemWContext>
    <StyledDropDownItemWContext
      className={false}
      key="b"
      label="B"
      onClick={[Function]}
      onMouseOver={[Function]}
      value="b"
    >
      B
    </StyledDropDownItemWContext>
    <StyledDropDownItemWContext
      className={false}
      key="c"
      label="C"
      onClick={[Function]}
      onMouseOver={[Function]}
      value="c"
    >
      C
    </StyledDropDownItemWContext>
  </DropDownListWContext>
</div>
`;
